File: petalinux/pkg/pacman-server/files/pacman_message.py
#!/usr/bin/env python3
"""
PACMAN message packing/unpacking utilities (low-level).
Header: 192-bit (24 bytes)
"""

import struct

# -----------------------------
# Message version
# -----------------------------
MSG_MAJOR_VERSION = 1
MSG_MINOR_VERSION = 0

# -----------------------------
# Message type constants
# -----------------------------
MSG_TYPE_REQ  = b'?'
MSG_TYPE_REP  = b'!'
MSG_TYPE_DATA = b'D'

MSG_TYPE_TABLE = {
    'REQ': MSG_TYPE_REQ,
    'REP': MSG_TYPE_REP,
    'DATA': MSG_TYPE_DATA
}

MSG_TYPE_TABLE_INV = {v:k for k,v in MSG_TYPE_TABLE.items()}

# -----------------------------
# Word type constants
# -----------------------------
WORD_TYPE_PING  = b'P'
WORD_TYPE_READ  = b'R'
WORD_TYPE_WRITE = b'W'
WORD_TYPE_DATA  = b'D'
WORD_TYPE_SYNC  = b'S'
WORD_TYPE_TRIG  = b'T'
WORD_TYPE_ERR   = b'E'

WORD_TYPE_TABLE = {
    'PING':  WORD_TYPE_PING,
    'READ':  WORD_TYPE_READ,    
    'WRITE': WORD_TYPE_WRITE,
    'DATA':  WORD_TYPE_DATA,
    'SYNC':  WORD_TYPE_SYNC,    
    'TRIG':  WORD_TYPE_TRIG,
    'ERR':   WORD_TYPE_ERR
}

WORD_TYPE_TABLE_INV = {v:k for k,v in WORD_TYPE_TABLE.items()}

# ----------------------------------------
# Word struct formats (192-bit / 24 bytes)
# ----------------------------------------

# PING:   0xWW000000 00000000 00000000 00000000 00000000 00000000
#    W=word type
# READ:   0xWWPP0000 00000000 AAAAAAAA RRRRRRRR 00000000 00000000 
#    W=word type, P=PACMAN id, A=address, R=value, S=Status
# WRITE:  0xWWPP0000 00000000 AAAAAAAA RRRRRRRR 00000000 00000000
#    W=word type, P=PACMAN id, A=address, R=value, S=Status
# DATA:   0xWWPPUUUU 00000000 TTTTTTTT TTTTTTTT DDDDDDDD DDDDDDDD
#    W=word type, P=PACMAN id, U=channel, T=timestamp, D=payload
# SYNC:   0xWWPPBBCC 00000000 TTTTTTTT TTTTTTTT SSSSSSSS 00000000
#    W=word type, P=PACMAN id, B=subtype, C=clock source, T=timestamp, S=Status
# TRIG:   0xWWPPBBGG 00000000 TTTTTTTT TTTTTTTT 00000000 00000000
#    W=word type, P=PACMAN id, B=subtype, G=trigger source, T=timestamp
# ERR:    0xWWPP0000 00000000 TTTTTTTT TTTTTTTT EEEEEEEE 00000000
#    W=word type, P=PACMAN id, T=timestamp, E=error code

WORD_LEN   = 24  # 192-bit
WORD_STRUCT_TABLE = {
    'PING':  struct.Struct('<c23x'),       # word_type
    'READ':  struct.Struct('<cB6xII8x'),   # word_type, pacman, address, value
    'WRITE': struct.Struct('<cB6xII8x'),   # word_type, pacman, address, value
    'DATA':  struct.Struct('<cBH4xQQ'),    # word_type, pacman, uart_channel, timestamp, payload
    'SYNC':  struct.Struct('<cBBB4xQI4x'), # word_type, pacman, sync_type, clock_source, timestamp, status  
    'TRIG':  struct.Struct('<cBBB4xQ8x'),  # word_type, pacman, trigger_type, trigger_source, timestamp
    'ERR':   struct.Struct('<cB6xQI4x')    # word_type, pacman, timestamp, error code
}

WORD_FIELD_TABLE = {
    'PING':  ("word_type",),
    'READ':  ("word_type", "pacman", "addr", "value"),
    'WRITE': ("word_type", "pacman", "addr", "value"),
    'DATA':  ("word_type", "pacman", "chan", "timestamp", "payload"),
    'SYNC':  ("word_type", "pacman", "sync_type", "clk_src", "timestamp", "status"),
    'TRIG':  ("word_type", "pacman", "trig_type", "trig_src", "timestamp"),
    'ERR':   ("word_type", "pacman", "timestamp", "error_code")
}

# -----------------------------
# Header struct (192-bit / 24 bytes)
# -----------------------------

# HEADER: 0xMMBBVVVV NNNNNNNN TTTTTTTT TTTTTTTT 00000000 00000000
#    M=messsage type, B=minor version, V=major version, N=number of words, T=timestamp

HEADER_STRUCT = struct.Struct('<cBHIQ8x') # message_type, 
HEADER_FIELDS = ("msg_type", "minor_version", "major_version", "n_words", "timestamp")
HEADER_LEN = HEADER_STRUCT.size

# -----------------------------
# Header functions
# -----------------------------
def pack_header(msg_type, n_words, timestamp):
    msg_type_byte = MSG_TYPE_TABLE[msg_type]
    return HEADER_STRUCT.pack(msg_type_byte, MSG_MINOR_VERSION, MSG_MAJOR_VERSION, n_words, timestamp)

def unpack_header(header_bytes):
    msg_type = MSG_TYPE_TABLE_INV[header_bytes[0:1]]
    values = HEADER_STRUCT.unpack(header_bytes)[1:]
    return (msg_type,) + values

def parse_header(header):
    return dict(zip(HEADER_FIELDS, header))

# -----------------------------
# Word functions
# -----------------------------
def pack_word(word_type, *data):
    word_type_byte = WORD_TYPE_TABLE[word_type]
    return WORD_STRUCT_TABLE[word_type].pack(word_type_byte, *data)

def unpack_word(word_bytes):
    word_type = WORD_TYPE_TABLE_INV[word_bytes[0:1]]
    values = WORD_STRUCT_TABLE[word_type].unpack(word_bytes)[1:]
    return (word_type,) + values

def parse_word(word):
    return dict(zip(WORD_FIELD_TABLE[word[0]], word))
    
# -----------------------------
# Message functions
# -----------------------------
def pack_msg(msg_type, msg_words, timestamp):
    n_words = len(msg_words)
    header_bytes = pack_header(msg_type, n_words, timestamp)
    body_bytes = b''.join([pack_word(*w) for w in msg_words])
    return header_bytes + body_bytes

def unpack_msg(msg_bytes):
    header = unpack_header(msg_bytes[:HEADER_LEN])
    words = []
    for i in range(HEADER_LEN, len(msg_bytes), WORD_LEN):
        words.append(unpack_word(msg_bytes[i:i+WORD_LEN]))
    return header, words

def content_ping():
    return ('PING',)

def content_read(*, addr, value=0, pacman=0):
    return ('READ',  pacman, addr, value)

def content_write(*, addr, value, pacman=0):
    return ('WRITE', pacman, addr, value)

def content_data(*, channel, timestamp, payload, pacman=0):
    return ('DATA', pacman, channel, timestamp, payload)

def content_sync(*, sync_type, timestamp, pacman=0, clock_source=0, status=0):
    return ('SYNC', pacman, sync_type, clock_source, timestamp, status)

def content_trig(*, trig_type, timestamp, pacman=0, trig_source=0):
    return ('TRIG', pacman, trig_type, trig_source, timestamp)

def content_err(*, error,  pacman=0, timestamp=0):
    return ('ERR', pacman, timestamp, error)


def print_header(header):
    parsed = parse_header(header)
    print("msg_type: {msg_type} version: {major_version}.{minor_version} n_words: {n_words} timestamp: {timestamp}".format(**parsed))    
    return

def print_word(word):
    format_strings = {
        "PING": "PING",
        "READ":  "word_type:  READ:  pacman: {pacman:04d} addr: 0x{addr:08x} value=0x{value:08x} ({value})",
        "WRITE": "word_type:  WRITE: pacman: {pacman:04d} addr: 0x{addr:08x} value=0x{value:08x} ({value})",
        "DATA":  "word_type:  DATA:  pacman: {pacman:04d} chan={chan:04d} payload=0x{payload:08X} timestamp={timestamp}",
        "SYNC":  "word_type:  SYNC:  pacman: {pacman:04d} sync_type: {sync_type} timestamp={timestamp}",
        "TRIG":  "word_type:  TRIG:  pacman: {pacman:04d} trig_type: {trig_type} timestamp={timestamp}",
        "ERR":   "word_type:  ERR:   pacman: {pacman:04d} error=0x{error_code:08X} timestamp={timestamp}",
    }    
    fmt = format_strings.get(word[0], "unknown")
    parsed = parse_word(word)
    print(fmt.format(**parsed))

File: petalinux/pkg/pacman-apps/files/src/drivers-linux/pacman_message.cc
#include <cstdio>
#include "pacman_message.hh"
#include <inttypes.h>

// -----------------------------
// Allowed sets
// -----------------------------
static const char valid_msg_types[] = {
  MSG_TYPE_REQ, MSG_TYPE_REP, MSG_TYPE_DATA, MSG_TYPE_STRING
};

static const char valid_word_types[] = {
  WORD_TYPE_PING, WORD_TYPE_READ, WORD_TYPE_WRITE, WORD_TYPE_DATA,
  WORD_TYPE_SYNC, WORD_TYPE_TRIG, WORD_TYPE_ERR
};

static bool is_valid_msg_type(char mt) {
    for (char c : valid_msg_types) {
        if (c == mt) return true;
    }
    return false;
}

static bool is_valid_word_type(char wt) {
    for (char c : valid_word_types) {
        if (c == wt) return true;
    }
    return false;
}

// check_msg: validate a message
bool check_msg(const pacman_msg_t* msg) {
    if (!msg) {
        printf("ERROR: msg pointer is NULL\n");
        return false;
    }

    uint16_t n_bytes = msg->header.n_bytes;

    // --- validate header msg_type ---
    char mt = msg->header.msg_type;
    if (not is_valid_msg_type(mt)){
      printf("ERROR: Unknown msg_type '%c' in header\n", mt);
      return false;
    }

    if (mt == MSG_TYPE_STRING) {
      // For strings, no per-word check is needed
      if (n_bytes > MAX_WORDS * WORD_BYTES) {
	printf("ERROR: string length %u exceeds maximum\n", n_bytes);
	return false;
      }
      return true;
    }


    uint16_t n_words = msg->header.n_bytes / WORD_BYTES;
    // Sanity check number of words
    if (n_words > MAX_WORDS) {
        printf("ERROR: n_bytes=%u exceeds MAX_WORDS=%u\n", n_bytes, MAX_WORDS);
        return false;
    }

    // Validate each word
    for (uint16_t i = 0; i < n_words; ++i) {
        const pacman_word_t* word = &msg->words[i];

        char wt = word->raw[0];  // first byte is word_type
        if (!is_valid_word_type(wt)) {
            printf("ERROR: Unknown word_type '%c' at index %u\n", wt, i);
            return false;
        }

        // Optional: add range checks per field if desired
        // e.g., pacman <= 255, addresses within 32-bit range, etc.
    }

    return true;
}

void print_header(const pacman_header_t* header, const char * prefix = "") {
  printf("version %d.%d n_bytes: %d ",header->version_major, header->version_minor, header->n_bytes);
  printf("timestamp:  %" PRIu64 "\n", header->timestamp);
}

void print_word(const pacman_word_t* word, const char * prefix = "") {
  char wt = word->raw[0];  // first byte is word_type

  switch (wt) {
  case WORD_TYPE_PING:
    printf("type: ping  ");
    break;
  case WORD_TYPE_READ:
    printf("type: read  pacman: %03d addr: 0x%04X value: 0x%08X", word->read.pacman, word->read.addr, word->read.value);
    break;
  case WORD_TYPE_WRITE:
    printf("type: read  pacman: %03d addr: 0x%04X value: 0x%08X", word->write.pacman, word->write.addr, word->write.value);
    break;
  case WORD_TYPE_DATA:
    printf("type: data  pacman: %03d chan: %5d payload:  0x%08x%08x timestamp: 0x%08x%08x", word->data.pacman,  word->data.chan,
	   upper_32(word->data.payload), lower_32(word->data.payload), upper_32(word->data.timestamp), lower_32(word->data.timestamp));
    break;
  case WORD_TYPE_SYNC:
    printf("type: sync  pacman %03d type: %c src: %d timestamp: 0x%08x%08x status: 0x%08x", word->sync.pacman, word->sync.sync_type,
	   word->sync.clk_src, upper_32(word->data.timestamp), lower_32(word->data.timestamp), word->sync.status);
    break;
  case WORD_TYPE_TRIG:
    printf("type: trig  pacman %03d type: %d src: %d timestamp: 0x%08x%08x", word->trig.pacman, word->trig.trig_type, word->trig.trig_src,
	    upper_32(word->data.timestamp), lower_32(word->data.timestamp));
    break;
  case WORD_TYPE_ERR:
    printf("type: err   pacman %03d timestamp: 0x%08x%08x error_code: 0x%08x", word->err.pacman,
	   upper_32(word->data.timestamp), lower_32(word->data.timestamp), word->err.error_code);
    break;
  default:
    printf("unknown ");
  }
  printf("\n");
}


void print_msg(const pacman_msg_t* msg, const char * prefix) {
  printf("%sheader:  ",prefix);
  print_header(&msg->header);

  if (msg->header.msg_type == MSG_TYPE_STRING) {
    printf("%sstring payload (%u bytes): ", prefix, msg->header.n_bytes);
    for (uint32_t i = 0; i < msg->header.n_bytes; ++i) {
      putchar(msg->raw[i]);
    }
    putchar('\n');
    return;
  }


  uint32_t n_words = msg->header.n_bytes / WORD_BYTES;
  for (uint32_t i = 0; i < n_words; ++i) {
    printf("%sword %3u: ", prefix, i);
    print_word(&msg->words[i], prefix);
  }
}
